node/actors/bft/src/testonly/twins/mod.rs

```diff
@@ -10,125 +10,45 @@
-                items,
-                output: Vec::new(),
-                acc,
-            };
-            // Generate all combinations
-            ps.go(0, 0);
-            // Take the results
-            ps.output
-        }
+    fn key(&self) -> &Self::Key {
+        (*self).key()
     }
+}
 
-    /// Recursively generate partitions.
-    ///
-    /// The algorithm goes item-by-item and has two indexes depending on what has happened
-    /// above it: the minimum and the maximum partition that it can (or has to) insert into.
-    ///
-    /// Take the table in [partitions] as an example with 4 items and 3 partitions.
-    /// * `A` is the first item, so it can only go into `P1` - `P2` can't be used
-    /// while `P1` is empty, otherwise we'd be generating redundant combinations. After that,
-    /// * `B` can go into minimally into `P1` because there are 2 more items after it, which
-    /// is enough to fill all remaining partitions; or it can go into `P2`, because `P1` is filled.
-    /// * `C` depends on what we did with `B`: if `B` is in `P1` then `C` has to minimally go into
-    /// `P2` to make sure no partition will be left empty at the end; if `B` is in `P2` then `C`
-    /// can go either in `P1`, `P2` or `P3`
+/// Abstraction for nodes which can have a twin.
+pub trait Twin: HasKey
+where
+    Self: PartialEq,
+{
+    /// Create a twin for the node.
+    fn to_twin(&self) -> Self;
+
+    /// Check if the other node is a twin of this one.
     ///
-    /// The algorithm has to traverse all possible placements with backtracking.
-    fn go(&mut self, idx: usize, first_empty: usize) {
-        // If we're beyond the last item, emit the complete combination.
-        if idx == self.num_items {
-            self.output.push(self.acc.clone());
-            return;
-        }
-        // Number of remaining items, including this one.
-        let rem_items = self.num_items - idx;
-        // Number of remaining empty partitions.
-        let rem_empty = self.num_partitions.saturating_sub(first_empty);
-        // Index of the last partition.
-        let last_part = self.num_partitions - 1;
-        // The maximum partition we can put into is the first empty one.
-        let max_part = min(first_empty, last_part);
-        // The minimum partition we have to put the item into is the one that still allows the remaining empty ones to be filled.
-        let min_part = if rem_empty == rem_items { max_part } else { 0 };
-        // Now generate all possible allocations.
-        for ins_part in min_part..=max_part {
-            self.acc[ins_part].push(self.items[idx]);
-            self.go(
-                idx + 1,
-                if ins_part == first_empty {
-                    first_empty + 1
-                } else {
-                    first_empty
-                },
-            );
-            // Backtrack
-            self.acc[ins_part].pop();
-        }
+    /// This is a reflexive, they are twins of each other.
```

Comment:
afaict this is a symmetric relation, but not reflexive.

Reply:
Ah yes, very true, thanks 🤦